### PR TITLE
feat: Add SwingSet configuration to purge vstorage within (re-)bootstrap

### LINKED ISSUE #7681: Bulldozer upgrade should purge non-exported vstorage entries
## What is the Problem Being Solved?

We should discard all pre-bulldozer vstorage data except for explicitly migrated entries, to avoid old vstorage remnants influencing the behavior of contracts or external consumers post-upgrade.

## Description of the Design

Extend SwingSet configuration to support specifying a list of subtree roots to purge, analogous to the `exportStorageSubtrees` configuration specifying a list of subtree roots to export to bootstrap.

## Security Considerations

We must be careful to not purge data _after_ bulldozer, but if we do then it should be possible to recover from an old vstorage snapshot.

## Scaling Considerations

We explicitly don't care about bulldozer bootstrap being slow.

## Test Plan

This should follow the same re-bootstrap pattern established by #7599.

### PR DESCRIPTION
**Pull request by gibson042**
Fixes #7681

## Description

Adds a `clearStorageSubtrees` configuration option of the same shape as `exportStorageSubtrees`. In case of overlap, the latter wins.

### Security Considerations

We must be careful to not purge data _after_ bulldozer, but if we do then it should be possible to recover from an old vstorage snapshot.

### Scaling Considerations

We explicitly don't care about bulldozer bootstrap being slow.

### Documentation Considerations

Covered with JSDoc.

### Testing Considerations

Covered in the test-vaults-upgrade bootstrap test.

### CODE REVIEW

**Comment by dckc** [comment]
> In case of overlap, the latter wins.

I expect we will want to export a bunch of stuff and then clear everything -- or at least: clear much of what was exported. Is that supported?

**Comment by gibson042** [comment]
> > In case of overlap, the latter wins.
> 
> I expect we will want to export a bunch of stuff and then clear everything -- or at least: clear much of what was exported. Is that supported?

No, but the recipient of the data has the paths to do so.

File: packages/SwingSet/src/types-external.js [hunk 1]
```diff
@@ -177,6 +177,9 @@ export {};
  * @typedef {object} SwingSetOptions
  * @property {string} [bootstrap]
  * @property {ConfigProposal[]} [coreProposals]
+ * @property {string[]} [clearStorageSubtrees] chain storage paths identifying roots of subtrees
+ *   for which data should be deleted (except for overlaps with exportStorageSubtrees, which
+ *   are preserved).
```
Comment (FUDCo) [inline packages/SwingSet/src/types-external.js]: Why is this here?  Things that are explicitly chain related do not belong in SwingSet.  These three options (`coreProposals`, `clearStorageSubtrees`, and `exportStorageSubtrees`) should not be in this package anywhere.

**Review by FUDCo — CHANGES_REQUESTED** [review]
The non-SwingSet-related config things in the types should go someplace else.

File: packages/SwingSet/src/types-external.js (hunk 1 above)
Comment (warner) [reply packages/SwingSet/src/types-external.js]: I think we've got some bleedover/(ab)use of the SwingSet `config` object and the thing of the same name in cosmic-swingset that includes additional non-SwingSet properties. `coreProposals` was the first, and these other two are following the same pattern, but I agree: cosmic-swingset should have a separate type that includes all three, so they won't be in the swingset package.

Comment (gibson042) [reply packages/SwingSet/src/types-external.js]: Hmm, that's a good point. Updated.

**Review by FUDCo — APPROVED** [review]
C'est bon.

File: packages/cosmic-swingset/src/launch-chain.js [hunk 2]
```diff
@@ -69,11 +83,13 @@ export async function buildSwingset(
   { debugName = undefined, slogCallbacks, slogSender },
 ) {
   const debugPrefix = debugName === undefined ? '' : `${debugName}:`;
-  /** @type {import('@agoric/swingset-vat').SwingSetConfig | null} */
-  let config = await loadSwingsetConfigFile(vatconfig);
-  if (config === null) {
-    config = loadBasedir(vatconfig);
+  let swingsetConfig = await loadSwingsetConfigFile(vatconfig);
```
Comment (warner) [inline packages/cosmic-swingset/src/launch-chain.js]: might be appropriate to call this one `config` as the union one, then define `swingSetConfig` as the leftovers once you remove the three `CosmicSwingsetConfig` properties, and pass `swingSetConfig` into swingset so it never sees the extras

File: packages/cosmic-swingset/src/launch-chain.js [hunk 3]
```diff
@@ -106,7 +122,11 @@ export async function buildSwingset(
       return;
     }
     if (!config) throw Fail`config not yet set`;
-    const { coreProposals, exportStorageSubtrees } = config;
+    const {
+      coreProposals,
+      clearStorageSubtrees,
+      exportStorageSubtrees = [],
```
Comment (warner) [inline packages/cosmic-swingset/src/launch-chain.js:125]: so `...swingSetConfig` here, and we change which one we pass into `initializeSwingSet` at line 218

**Review by warner — CHANGES_REQUESTED** [review]
small cleanups to make

File: packages/cosmic-swingset/src/launch-chain.js (hunk 2 above)
Comment (warner) [reply packages/cosmic-swingset/src/launch-chain.js]: actually given the number of references to `config`, maybe this combined one should be called `combinedConfig`, and then we quickly split it into `{ coreProposals, clearStorageSubstrees, exportStorageSubtrees, ...config }`

**Review by warner — APPROVED** [review]
LGTM, thanks for the cleanup
